curvefs/src/metaserver/transaction.cpp

```diff
@@ -29,84 +29,101 @@ namespace metaserver {
 }
 
-bool RenameTx::Rollback() {
-    FOR_EACH_DENTRY(ROLLBACK);
-    return true;
+bool RenameTx::Rollback(std::int64_t logIndex) {
+    return storage_->RollbackTx(dentrys_, logIndex) == MetaStatusCode::OK;
 }
 
-uint64_t RenameTx::GetTxId() {
-    return txId_;
-}
+uint64_t RenameTx::GetTxId() { return txId_; }
 
-uint64_t RenameTx::GetTxSequence() {
-    return txSequence_;
-}
+uint64_t RenameTx::GetTxSequence() { return txSequence_; }
 
-std::vector<Dentry>* RenameTx::GetDentrys() {
+std::vector<Dentry> *RenameTx::GetDentrys() { return &dentrys_; }
+
+const std::vector<Dentry> *RenameTx::GetConstDentrys() const {
     return &dentrys_;
 }
 
-inline bool RenameTx::operator==(const RenameTx& rhs) {
+inline bool RenameTx::operator==(const RenameTx &rhs) {
     return dentrys_ == rhs.dentrys_;
 }
 
-std::ostream& operator<<(std::ostream& os, const RenameTx& renameTx) {
+std::ostream &operator<<(std::ostream &os, const RenameTx &renameTx) {
     auto dentrys = renameTx.dentrys_;
     os << "txId = " << renameTx.txId_;
     for (size_t i = 0; i < dentrys.size(); i++) {
-        os << ", dentry[" << i << "] = ("
-           << dentrys[i].ShortDebugString() << ")";
+        os << ", dentry[" << i << "] = (" << dentrys[i].ShortDebugString()
+           << ")";
     }
     return os;
 }
 
-TxManager::TxManager(std::shared_ptr<DentryStorage> storage)
-    : storage_(storage) {}
+TxManager::TxManager(std::shared_ptr<DentryStorage> storage,
+                     common::PartitionInfo partitionInfo)
+    : storage_(std::move(storage)), conv_(std::make_shared<Converter>()),
+      partitionInfo_(std::move(partitionInfo)) {}
 
-MetaStatusCode TxManager::PreCheck(const std::vector<Dentry>& dentrys) {
+MetaStatusCode TxManager::PreCheck(const std::vector<Dentry> &dentrys) {
     auto size = dentrys.size();
     if (size != 1 && size != 2) {
         return MetaStatusCode::PARAM_ERROR;
     } else if (size == 2) {
-         if (dentrys[0].fsid() != dentrys[1].fsid() ||
-             dentrys[0].txid() != dentrys[1].txid() ||
-             dentrys[0].txsequence() != dentrys[1].txsequence()) {
+        if (dentrys[0].fsid() != dentrys[1].fsid() ||
+            dentrys[0].txid() != dentrys[1].txid() ||
+            dentrys[0].txsequence() != dentrys[1].txsequence()) {
             return MetaStatusCode::PARAM_ERROR;
         }
     }
 
     return MetaStatusCode::OK;
 }
 
-MetaStatusCode TxManager::HandleRenameTx(const std::vector<Dentry>& dentrys) {
+void TxManager::SerializeRenameTx(const RenameTx &in,
+                                  PrepareRenameTxRequest *out) {
+    auto *dentrys = in.GetConstDentrys();
+    out->set_poolid(partitionInfo_.poolid());
+    out->set_copysetid(partitionInfo_.copysetid());
+    out->set_partitionid(partitionInfo_.partitionid());
+    *out->mutable_dentrys() = {dentrys->begin(), dentrys->end()};
+}
+
+void TxManager::LoadPengingTx() {
+    metaserver::TransactionRequest request;
+    if (storage_->GetPendingTx(&request)) {
```

Comment:
`GetPendingTx` may return false, is it safe to ignore this error?

Reply:
Yes, if the snapshot generated by old version of metaserver, there are no transaction in snapshot.

Reply:
Or we may meet a disk error, in that case nothing we can do, and the snapshot files maybe broken.

Reply:
I have expose error to `MetaStore::Load()`.